aspnetcore/performance/performance-best-practices.md

```diff
@@ -132,19 +131,200 @@ Recommendations:
+-->
+`HttpContext` is only valid as long as there is an active HTTP request in the ASP.NET Core pipeline. The entire ASP.NET Core pipeline is an asynchronous chain of delegates that executes every request. When the `Task` returned from this chain completes, the `HttpContext` is recycled.
+
+**Do not do this:** The following example uses `async void`:
+
+* Which is **ALWAYS** a bad practice in ASP.NET Core apps.
+* Accesses the `HttpResponse` after the HTTP request is complete.
+* Crashes the process.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs?name=snippet1)]
+
+**Do this:** The following example returns a `Task` to the framework so the HTTP request doesn't complete until the action completes.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/AsyncSecondController.cs?name=snippet1)]
+
+## Do not capture the HttpContext in background threads
+
+**Do not do this:** The following example shows a closure is capturing the `HttpContext` from the `Controller` property. This is a bad practice because the work item could:
+
+* Run outside of the request scope.
+* Attempt to read the wrong `HttpContext`.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetFirstController.cs?name=snippet1)]
+
+**Do this:** The following example:
+
+* Copies the data required in the background task during the request.
+* Doesn't reference anything from the controller.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetFirstController.cs?name=snippet2)]
+
+## Do not capture services injected into the controllers on background threads
+
+**Do not do this:** The following example shows a closure is capturing the `DbContext` from the `Controller` action parameter. This is a bad practice.  The work item could run outside of the request scope. The `PokemonDbContext` is scoped to the request, resulting in an `ObjectDisposedException`.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet1)]
+
+**Do this:** The following example:
+
+* Injects an <xref:Microsoft.Extensions.DependencyInjection.IServiceScopeFactory> in order to create a scope in the background work item. `IServiceScopeFactory` is a singleton.
+* Creates a new dependency injection scope in the background thread.
+* Doesn't reference anything from the controller.
+* Doesn't capture the `PokemonDbContext` from the incoming request.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2)]
+
+The following highlighted code:
+
+* Creates a scope for the lifetime of the background operation and resolves services from it.
+* Uses `PokemonDbContext` from the correct scope.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2&highlight=9-16)]
+
+## Avoid adding headers after the HttpResponse has started
```

Comment:
```suggestion
## Do not modify the status code or headers after the response body has started
```

Reply:
Related: Do not call `next()` if you have already started writing to the response body. Components only expect to be called if it's possible for them to handle/manipulate the response.